### PR TITLE
Add ewm pandas method to dask

### PR DESCRIPTION
**Pull request by psimaj**
This draft pull request contains a work-in-progress implementation of the ewm computations from pandas, as requested in #4184.
At this point I've implemented a framework based on Rolling that should allow me to implement the `mean`, `var` and `std` methods from pandas - I'm not quite sure about handling `cov` and `corr` just yet. The implementation is based on an assumption that in order to compute values in the current partition, we only need to know the last computed value in the previous partition - this holds for the three afromentioned methods due to their recursive nature. While this approach doesn't have a trivial parallelization, it allows for linear complexity, so it can potentially be very efficient.
Note that the implementation is quite rough at this point - I'm open to suggestions regarding both the approach and the implementation.

### CODE REVIEW

**Comment by jcrist** [comment]
Hi @psimaj, thanks for the PR. I suggest pushing on ahead with what you have, and asking questions as you get stuck. I wouldn't worry about `cov` and `corr` yet - those can come later if people need them. The general approach here looks pretty good though.

File: dask/dataframe/ewm.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        )
+
+    partial_keys = list(dsk)
+    dsk.update(
+        {
+            (name_merge, i): (
+                merge_func,
+                input_key,
+                partial_keys[:i],
+                df.__dask_keys__()[:i],
```
Comment (TomAugspurger) [inline dask/dataframe/ewm.py:58]: To make sure I understand, this won't (can't) be computed in parallel right? The results from partition `4` depend on the result from partitions `0, 1, 2, 3`?

Comment (TomAugspurger) [reply dask/dataframe/ewm.py:58]: And it's just the *result* of previous partitions that we need (at least for the functions you've done so far) Not the inputs?

Comment (psimaj) [reply dask/dataframe/ewm.py:58]: Yes, merging the results from partitions isn't parallelizable in this implementation. This is the cost of avoiding quadratic work (and linear time) that would be required to compute the EWM operations by definition.
And yes, that `input_key` does refer to a result from previous partitions, not the inputs. I will change the name to something that make more sense.

File: dask/dataframe/core.py [hunk 2]
```diff
@@ -1441,6 +1441,34 @@ def rolling(self, window, min_periods=None, center=False, win_type=None, axis=0)
...
+        return EWM(
+            self,
+            com=com,
+            span=span,
+            halflife=halflife,
+            alpha=alpha,
+            min_periods=min_periods,
+            adjust=adjust,
+            ignore_na=ignore_na,
+            axis=0,
```
[the remark this answers is not in the document]
Comment (psimaj) [reply dask/dataframe/core.py:1469]: Sorry, it's a typo. It will be changed to `axis=axis`.

**Comment by psimaj** [comment]
Thank you for the comments so far, I will be working on addressing them and implementing the remaining functionalities in the following 2 weeks, I'm planning to have it ready for an another review in about 10 days.

File: dask/dataframe/ewm.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    name = "{0}-{1}".format(func_name, token)
+    name_partial = "partial-" + func_name
+    name_merge = "merge-" + merge_func_name
+
+    dsk = {}
+
+    for i, input_key in enumerate(df.__dask_keys__()):
+        dsk.update({(name_partial, i): (func, input_key, args, kwargs)})
+        kwargs["min_periods"] = max(
+            0, kwargs["min_periods"] - df.partitions[i].shape[kwargs["axis"]]
```
[the remark this answers is not in the document]
Comment (psimaj) [reply dask/dataframe/ewm.py:48]: Honestly I can't see how this is handled in `Rolling`. 
It's meant to adjust the minimal allowed length of the window for the pandas method based on the parition the method will be invoked on, but now I see that doing it this way makes little sense, as there will be only one partition for which this parameter is relevant (all other partitions will either have it at 0, or at their length, in which case the operation won't be performed at all). I will change it to take that into account.

**Comment by quasiben** [comment]
It looks like this PR stalled out a bit. @michaelaye if you want to submit a new PR that would be great!

**Comment by MichaelKarpe** [comment]
> It looks like this PR stalled out a bit. @michaelaye if you want to submit a new PR that would be great!

@quasiben done with #6796!

**Comment by jsignell** [comment]
Closing this in favor of #6796.

**Pull request closed without merging**